### PR TITLE
Added ability to add additional headers and cookies to the GraphQL response

### PR DESCRIPTION
**Pull request by donbeave**
This PR is mostly related to that issue: https://github.com/micronaut-projects/micronaut-graphql/issues/98

Basically I just added a way to add headers and cookies inside GraphQL `DataFetcher` (graphql-java) or `GraphQLResolver` (graphql-java-tools).

I also added an example `jwt-security` which shows how to implement authentication by using GraphQL API.

I hope you like this idea.

### CODE REVIEW

**Comment by donbeave** [comment]
Sorry for pushing you guys, @marceloverdijk @sdelamo @graemerocher @ilopmar , do you have time to take a look at this PR? Is there anything that I need to fix or improve?

**Comment by marceloverdijk** [comment]
Hi @donbeave 

The idea of the `GraphQLExecutionInputCustomizer` is so everybody can provide a custom impl instead of the ` DefaultGraphQLExecutionInputCustomizer` to e.g. push something to the GraphQL context (like you are doing).

But I'm not sure about the `GraphQLContextHttpUtils `.

I haven't touched Micronaut lately so I cannot recall it from my head, but is there no other way in Micronaut to get the HTTP request. I would prefer such an solution instead of cluttering the GraphQL context for everybody automatically.

**Comment by donbeave** [comment]
Hi, @marceloverdijk

The main problem actually is not getting an HTTP request, probably there is a way to get an HTTP request from some service or holder, like how Micronaut Security works. But I can not find any way to manipulate response headers (include cookies), that's why eventually I have added `GraphQLContextHttpUtils` class. I'm not really like this implementation, because everybody can change their GraphQL context with any other object, which will completely destroy this way of transmission response headers from GraphQL data-fetchers/resolvers. I personally more like the idea of using the original `GraphQLContext` (https://github.com/graphql-java/graphql-java/blob/8b22a39f358e32265b8f89eec176712490db7805/src/main/java/graphql/GraphQLContext.java), without replacing it with some custom class, if need to add any additional information to the context, better use GraphQLExecutionInputCustomizer without replacing the original context instance, but only modify values inside that object.

Regarding `GraphQLContextHttpUtils`, this is only a utility class, which can modify the original `GraphQLContext` instance by adding some additional information (like headers) to the response. I can not find any other way to do it, there is only `DataFetchingEnvironment` (https://github.com/graphql-java/graphql-java/blob/8b22a39f358e32265b8f89eec176712490db7805/src/main/java/graphql/schema/DataFetchingEnvironment.java) class that can be used in data fetcher class, inside this class the only context looks suitable for passing information to the controller. So just because the response object is creating inside `GraphQLController`, that's why I need to use some utility class to pass additional headers through GraphQL context.

Maybe if you have idea how can make it better, I'm ok to implement it.

**Comment by donbeave** [comment]
@marceloverdijk what do you think? Sorry for pushing, I'm just migrating current Spring Boot GraphQL application to Micronaut and without such changes, there is no way to implement authentication inside GraphQL API.

**Comment by graemerocher** [comment]
@donbeave @marceloverdijk Does it make sense to add the security functionality to a separate module? Or use events somehow to propagate state?

**Comment by donbeave** [comment]
@graemerocher maybe yes, but my example not only shows security things, but also rate-limiting by using headers. Because GraphQL can be a gateway for multiple (for example gRPC) microservices, each GraphQL field can return custom headers with rate limit information. This makes sense to have an ability to read request headers from GraphQL resolvers and also have an ability to manipulate response headers.

**Comment by marceloverdijk** [comment]
I fully agree it should be easy to get request header and manipulate response header/cookies from from GraphQL resolvers/data fetchers.

@graemerocher are there any holders to do this, especially setting the response headers?

@donbeave with the `GraphQLExecutionInputCustomizer` you can get the current context and modify it without replacing.
Something like:

````
customize(graphql.ExecutionInput executionInput, io.micronaut.http.HttpRequest httpRequest) {
    Object context = executionInput.getContext();
    // cast and modify
}
````

**Comment by donbeave** [comment]
@graemerocher sorry for being so annoying, regarding to that question: 

> are there any holders to do this, especially setting the response headers?

maybe you can give me some direction on what to check?

**Comment by graemerocher** [comment]
@marceloverdijk there is no holder to get the response no, it would have to be passed down the chain

**Comment by donbeave** [comment]
@marceloverdijk seems only two options,
1) use static methods (like I did), which will work with GraphQLContext original instance
2) extend original GraphQL context and add these methods to work with request and responce headers

What do you think?

**Comment by donbeave** [comment]
@marceloverdijk sorry for pushing again, I'm just stuck here in waiting for this feature. I understand you are busy now and didn't touch Micronaut lately, maybe I can help somehow with this project?

**Comment by marceloverdijk** [comment]
@donbeave np, I'm only a bit short in time at the moment.

I think your main problem is being able to set the response headers, right?

You should be able to set the request in the graphql context with a custom `GraphQLExecutionInputCustomizer` which calls your utility class.

**Comment by donbeave** [comment]
@marceloverdijk 

> I think your main problem is being able to set the response headers, right?

Actually two problems:
1) set response headers
2) get request headers

> You should be able to set the request in the graphql context with a custom GraphQLExecutionInputCustomizer which calls your utility class.

It only solves the second problem, which is getting request headers. But with `HttpRequest` instance I can not modify response headers. Response headers can be added only in `GraphQLController`. The current implementation even is not returning `HttpResponse` entity, only `String`, which in result will be without any additional headers.

Maybe do it in another way, how about change this code in controller to:

```
private Publisher<MutableHttpResponse<String>> executeRequest(
            String query,
            String operationName,
            Map<String, Object> variables,
            HttpRequest httpRequest) {
        GraphQLInvocationData invocationData = new GraphQLInvocationData(query, operationName, variables);
        // create empty response entity first and pass it to GraphQLInvocation
        MutableHttpResponse<String> response = HttpResponse.status(HttpStatus.OK);
        Publisher<ExecutionResult> executionResult = graphQLInvocation.invoke(invocationData, httpRequest, response);
        Publisher<GraphQLResponseBody> responseBody = graphQLExecutionResultHandler.handleExecutionResult(executionResult);
        return Publishers.map(responseBody, graphQLResponseBody -> httpResponse.body(graphQLJsonSerializer.serialize(responseBody)));
    }
```

As well need to change `GraphQLInvocation` and `GraphQLExecutionInputCustomizer` interfaces:

```
public interface GraphQLInvocation {
    Publisher<GraphQLExecution> invoke(GraphQLInvocationData invocationData, HttpRequest httpRequest, MutableHttpResponse<String> httpResponse);
}

public interface GraphQLExecutionInputCustomizer {
    Publisher<ExecutionInput> customize(ExecutionInput executionInput, HttpRequest httpRequest, MutableHttpResponse<String> httpResponse);
}
```

With such changes, I can remove `GraphQLContextHttpUtils` from here and implement it in my project only, request, and response entities I can add to context by overriding `GraphQLExecutionInputCustomizer`. This will solve the first problem as well.

What I don't like in this solution, it gives too much flexibility to the developer. Because developer change modifies HttpResponse somewhere inside GraphQL data fetchers, for example by overriding response status, content length, content type, or content-encoding... That's why I prefer to have a utility class in the official library with fixed functionality.

**Comment by marceloverdijk** [comment]
I like the second approach to be honest. I think it gives all the flexibility we need.

**Comment by donbeave** [comment]
@marceloverdijk good, I will update this PR today

**Comment by marceloverdijk** [comment]
@donbeave 👍 for reviewing it might better to create a new PR?

**Comment by donbeave** [comment]
@marceloverdijk I think this one is ok, because it will shows only changes compare to the master code

**Comment by donbeave** [comment]
@marceloverdijk done, please check. I also replaced `javax.annotation.Nullable` with SpotBugs one due Micronaut 2 is migrated to SpotBugs.

**Comment by marceloverdijk** [comment]
Thx @donbeave I will try to have a look tomorrow.

**Comment by ilopmar** [comment]
@marceloverdijk please ping us before doing the release so we can coordinate with the pending PRs. Also keep in mind that this is a new minor version because it adds new features.

File: graphql/src/main/java/io/micronaut/configuration/graphql/DefaultGraphQLExecutionInputCustomizer.java [hunk 1]
```diff
@@ -35,7 +38,14 @@
      * {@inheritDoc}
      */
     @Override
-    public Publisher<ExecutionInput> customize(ExecutionInput executionInput, HttpRequest httpRequest) {
+    public Publisher<ExecutionInput> customize(ExecutionInput executionInput, HttpRequest httpRequest,
+                                               @Nullable MutableHttpResponse<String> httpResponse) {
+        // we store original request and response, since it might be needed inside data fetchers
+        if (executionInput.getContext() != null && executionInput.getContext() instanceof GraphQLContext) {
```
Comment (marceloverdijk) [inline graphql/src/main/java/io/micronaut/configuration/graphql/DefaultGraphQLExecutionInputCustomizer.java:44]: This is not needed. Users should create there own `GraphQLExecutionInputCustomizer` wehere they can choose whether to store the request or response.

File: graphql/src/main/java/io/micronaut/configuration/graphql/DefaultGraphQLInvocation.java [hunk 2]
```diff
@@ -35,9 +36,9 @@
  * @author Marcel Overdijk
  * @author Graeme Rocher
  * @author James Kleeh
- * @since 1.0
- * @see GraphQLExecutionInputCustomizer#customize(ExecutionInput, HttpRequest)
+ * @see GraphQLExecutionInputCustomizer#customize(ExecutionInput, HttpRequest, MutableHttpResponse)
  * @see GraphQL#executeAsync(ExecutionInput.Builder)
+ * @since 1.0
```
Comment (marceloverdijk) [inline graphql/src/main/java/io/micronaut/configuration/graphql/DefaultGraphQLInvocation.java:41]: Why is this moved below @see ?

File: graphql/src/main/java/io/micronaut/configuration/graphql/DefaultGraphQLInvocation.java [hunk 3]
```diff
@@ -75,7 +77,12 @@ public DefaultGraphQLInvocation(
             executionInputBuilder.dataLoaderRegistry(dataLoaderRegistry.get());
         }
         ExecutionInput executionInput = executionInputBuilder.build();
-        return Flowable.fromPublisher(graphQLExecutionInputCustomizer.customize(executionInput, httpRequest))
+
+        Flowable<ExecutionInput> executionInputFlowable = graphQLExecutionInputCustomizer != null
```
Comment (marceloverdijk) [inline graphql/src/main/java/io/micronaut/configuration/graphql/DefaultGraphQLInvocation.java:81]: Why is this added?  The `GraphQLExecutionInputCustomizer`  should never be `null` as we use a `DefaultGraphQLExecutionInputCustomizer` ? Did you encounter any issues with current approach?

File: graphql/src/main/java/io/micronaut/configuration/graphql/GraphQLContextKeys.java [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+ */
+package io.micronaut.configuration.graphql;
+
+/**
+ * Constant definitions for the standard keys used by this plugin in
+ * {@link graphql.GraphQLContext} or {@link io.micronaut.websocket.WebSocketSession}.
+ *
+ * @since 2.1.0
+ */
+public final class GraphQLContextKeys {
```
Comment (marceloverdijk) [inline graphql/src/main/java/io/micronaut/configuration/graphql/GraphQLContextKeys.java:24]: This class is not needed anymore imo.

**Comment by marceloverdijk** [comment]
Hi @donbeave thx for the changes. I have commented some parts in the PR could you pls have alook?
Also the `GraphQLWsController.java` contains references to the request key.

File: graphql/src/main/java/io/micronaut/configuration/graphql/DefaultGraphQLExecutionInputCustomizer.java (hunk 1 above)
Comment (donbeave) [reply graphql/src/main/java/io/micronaut/configuration/graphql/DefaultGraphQLExecutionInputCustomizer.java:44]: Ok, I'll remove it.

File: graphql/src/main/java/io/micronaut/configuration/graphql/DefaultGraphQLInvocation.java (hunk 2 above)
Comment (donbeave) [reply graphql/src/main/java/io/micronaut/configuration/graphql/DefaultGraphQLInvocation.java:41]: Sorry, it was IntelliJ IDEA auto-format.

File: graphql/src/main/java/io/micronaut/configuration/graphql/DefaultGraphQLInvocation.java (hunk 3 above)
Comment (donbeave) [reply graphql/src/main/java/io/micronaut/configuration/graphql/DefaultGraphQLInvocation.java:81]: Sometimes it can be null, I have encountered this issue. I thought it was a bug, but when I opened this file I saw the constructor allow null values here:

```java
    public DefaultGraphQLInvocation(
            GraphQL graphQL,
            @Nullable GraphQLExecutionInputCustomizer graphQLExecutionInputCustomizer,
            @Nullable Provider<DataLoaderRegistry> dataLoaderRegistry) {
```

That's why I have added that checking.

I can provide an example when `graphQLExecutionInputCustomizer` can be `null`.

File: graphql/src/main/java/io/micronaut/configuration/graphql/GraphQLContextKeys.java (hunk 4 above)
Comment (donbeave) [reply graphql/src/main/java/io/micronaut/configuration/graphql/GraphQLContextKeys.java:24]: Ok, I'll remove it.

File: graphql/src/main/java/io/micronaut/configuration/graphql/DefaultGraphQLExecutionInputCustomizer.java (hunk 1 above)
Comment (donbeave) [reply graphql/src/main/java/io/micronaut/configuration/graphql/DefaultGraphQLExecutionInputCustomizer.java:44]: Originally I think maybe many ppl like me don't want to copy-paste own version of `GraphQLExecutionInputCustomizer`, and prefer to have such feature out-of-box. Because in my apps, almost all of the using headers inside GraphQL data fetchers.

File: graphql/src/main/java/io/micronaut/configuration/graphql/DefaultGraphQLInvocation.java (hunk 3 above)
Comment (marceloverdijk) [reply graphql/src/main/java/io/micronaut/configuration/graphql/DefaultGraphQLInvocation.java:81]: Interesting. As the `DefaultGraphQLExecutionInputCustomizer` has `@Singleton` annotation it should always have that one, unless replaced by the user.
You don't have to work out a complete example, but it's worth investigating how it happened in your case.

Comment (donbeave) [reply graphql/src/main/java/io/micronaut/configuration/graphql/DefaultGraphQLInvocation.java:81]: Here is an example: https://github.com/donbeave/micronaut-graphql-customizer-null. I think if switch to use `Provider`, it will solve such error probably.

Comment (donbeave) [reply graphql/src/main/java/io/micronaut/configuration/graphql/DefaultGraphQLInvocation.java]: Yeah, using `Provider` can fix that NPE issue: https://github.com/micronaut-projects/micronaut-graphql/pull/99/commits/c0a0c4c66842e4c468896040c78ae19ebc2b1d91

**Comment by donbeave** [comment]
Hi, @marceloverdijk !

I made code review fixes, please check again.

**Comment by donbeave** [comment]
@marceloverdijk sorry for pinging you again )

File: graphql/src/test/groovy/io/micronaut/configuration/graphql/GraphQLControllerSpec.groovy [hunk 5]
```diff
@@ -68,6 +77,16 @@ class GraphQLControllerSpec extends Specification {
         executionInput = null
         1 * graphQL.executeAsync(_) >> { ExecutionInput executionInput ->
             this.executionInput = executionInput
+            if (executionInput.query == "{ testHeaders }") {
+                GraphQLContext graphQlContext = executionInput.getContext()
```
Comment (marceloverdijk) [inline graphql/src/test/groovy/io/micronaut/configuration/graphql/GraphQLControllerSpec.groovy:81]: I think these expectations/asserts should be moved to the particular test (so no need to test on the query "testHeaders" here .

File: graphql/src/main/java/io/micronaut/configuration/graphql/DefaultGraphQLInvocation.java (hunk 3 above)
Comment (marceloverdijk) [reply graphql/src/main/java/io/micronaut/configuration/graphql/DefaultGraphQLInvocation.java]: Did you find out why it was `null` without the provider, the code seemed to be ok so I wonder what was going on.
I understand `Provider` fixes it but imo it seems unnecessary as the Micronaut-GraphQL should always have a  `GraphQLExecutionInputCustomizer`; either a custom one or the `DefaultGraphQLExecutionInputCustomizer`.

**Comment by marceloverdijk** [comment]
@donbeave I reviewed the files and left 1 comment for the test.
I will also look into the need for the `Provider<GraphQLExecutionInputCustomizer>` as it should never be `null` imo. I will check out your repo to see.
And I want to perform some tests locally with the changes.
We are getting there ;-)

File: graphql/src/test/groovy/io/micronaut/configuration/graphql/GraphQLControllerSpec.groovy (hunk 5 above)
Comment (donbeave) [reply graphql/src/test/groovy/io/micronaut/configuration/graphql/GraphQLControllerSpec.groovy:81]: The problem here is this is in `setup` method already, I mean that assert:

```groovy
1 * graphQL.executeAsync(_) >> { ExecutionInput executionInput ->
```

I do not very understand how to move my additional changes away from here, because this `setup` method will be executed for each test, if I remove my changes from here, then I will get the error from my test:

```groovy
Too few invocations for:

1 * graphQL.executeAsync(_) >> { ExecutionInput executionInput ->
            this.executionInput = executionInput
            return executionResult
        }   (0 invocations)
```

Because that setup method will be executed for my test as well.

I can see a few options here:

1) Move my test to a separate class
2) Add checking for current test name in the setup method by using `specificationContext.currentIteration.name`, which is for me is a hack-way and doesn't seem to be a correct way
3) Rename `setup` method and for each, not my test add invoking that renamed method

What do you think, maybe you have a better suggestion?

File: graphql/src/main/java/io/micronaut/configuration/graphql/DefaultGraphQLInvocation.java (hunk 3 above)
Comment (donbeave) [reply graphql/src/main/java/io/micronaut/configuration/graphql/DefaultGraphQLInvocation.java]: `null` will be if there are multiple `Singleton` classes with the same interface `GraphQLExecutionInputCustomizer`, but none of them marked with `Primary` annotation, which is what my example project shows. Of course, this is a developer issue, I think, and the developer must add `Primary` annotation always in such cases, but some can forget (which was I did), and with NPE error thought it was not an issue in their code, but probably in the plugin code, that's why Provider looks better for me in such scenarios.

Comment (donbeave) [reply graphql/src/main/java/io/micronaut/configuration/graphql/DefaultGraphQLInvocation.java]: With using `Provider` the error will be much more clearer:

```
Caused by: io.micronaut.context.exceptions.NonUniqueBeanException: Multiple possible bean candidates found: [io.micronaut.configuration.graphql.DefaultGraphQLExecutionInputCustomizer, example.graphql.RequestResponseCustomizer]
```

Comment (marceloverdijk) [reply graphql/src/main/java/io/micronaut/configuration/graphql/DefaultGraphQLInvocation.java]: Interesting, you should indeed use @Replaces on the ` RequestResponseCustomizer` class.
Maybe this is something that can be improved in Micronaut in general to have clearer error message?
@graemerocher @ilopmar is it recommended to inject a `Provider<GraphQLExecutionInputCustomizer>`  even if we know the Micronaut-GraphQL extensions always supplies a `DefaultGraphQLExecutionInputCustomizer`? I don't think it's good to do this only for getting a better error message, so I would say we should keep it without `Provider`.

Comment (donbeave) [reply graphql/src/main/java/io/micronaut/configuration/graphql/DefaultGraphQLInvocation.java]: Agree, I also think it's better to have the same error with/or without `Provider`.

Comment (ilopmar) [reply graphql/src/main/java/io/micronaut/configuration/graphql/DefaultGraphQLInvocation.java]: `GraphQLExecutionInputCustomizer` shouldn't be `Nullable` if there is always an instance created by the library. The `package-info.java` file is annotated with `@RequiresGraphQL`, which requires a `GraphQL` bean and the property `graphql.enabled` not being `false`. Only in case that those conditions are false, the bean won't be created. But none of the beans in the package will be created either.

The library doesn't include a factory to create the `GraphQL` bean, and according to the docs, its seems the users need to do it. So again, I don't see how that bean can be `null` when the "requirements" for creating the beans are true.

Comment (marceloverdijk) [reply graphql/src/main/java/io/micronaut/configuration/graphql/DefaultGraphQLInvocation.java]: @ilopmar I think the issue @donbeave was experiencing when there are 2 `GraphQLExecutionInputCustomizer` beans. In that case the error was not clear. 
I think the right approach for now is to remove using the `Provider`  in this PR @donbeave .

Comment (ilopmar) [reply graphql/src/main/java/io/micronaut/configuration/graphql/DefaultGraphQLInvocation.java]: I see... When a user replaces a Micronaut bean it should add `@Primary` to use it as you both have been discussing. 
I've seen the "multiple beans candidates" some times and the error has been clear without using a `Provider`.

Btw, another option is to annotate the `DefaultGraphQLExecutionInputCustomizer` with `@Requires(missingBeans = {GraphQLExecutionInputCustomizer.class})` so it will only be created if the user doesn't provide its own bean.

Comment (donbeave) [reply graphql/src/main/java/io/micronaut/configuration/graphql/DefaultGraphQLInvocation.java]: The `null` value will be only if there 2 or more `GraphQLExecutionInputCustomizer` beans already and none of them marked with `Primary` annotation. If use provider than it will not be nullable, but instead when try to get bean it will throw a `NonUniqueBeanException` exception.

Ok, @marceloverdijk, I'll remove `Provider` from here.

Comment (marceloverdijk) [reply graphql/src/main/java/io/micronaut/configuration/graphql/DefaultGraphQLInvocation.java]: And probably a good addition what @ilopmar mentioned, that should also cause no duplicate `GraphQLExecutionInputCustomizer` beans.

Comment (ilopmar) [reply graphql/src/main/java/io/micronaut/configuration/graphql/DefaultGraphQLInvocation.java]: > The `null` value will be only if there 2 or more `GraphQLExecutionInputCustomizer` beans already and none of them marked with `Primary` annotation. If use provider than it will not be nullable, but instead when try to get bean it will throw a `NonUniqueBeanException` exception.

If you remove the `@Nullable` annotation from the constructor it will fail with the "multiple bean candidates" messages when there are two beans because Micronaut won't know which one to inject. And with this you won't need to add any additional `@Requires` or mark the bean as `@Primary`. Only the user will need to do it, but they will know because the error message will be clear.

Comment (donbeave) [reply graphql/src/main/java/io/micronaut/configuration/graphql/DefaultGraphQLInvocation.java]: @marceloverdijk I'll add it too.

Comment (donbeave) [reply graphql/src/main/java/io/micronaut/configuration/graphql/DefaultGraphQLInvocation.java]: Done here: https://github.com/micronaut-projects/micronaut-graphql/pull/99/commits/d74bc324f09a13b162ea9e43e920a914df17f6e0

Comment (donbeave) [reply graphql/src/main/java/io/micronaut/configuration/graphql/DefaultGraphQLInvocation.java]: @ilopmar got it, thank you for explanation how it works

**Review by marceloverdijk — APPROVED** [review]
@donbeave thx a lot for this PR, I will merge it.

**Comment by donbeave** [comment]
Since it's approved already, @marceloverdijk @ilopmar, can we merge it and build snapshot version?

**Comment by marceloverdijk** [comment]
@donbeave We are working on it, you need a little bit more of patience ;-). I can't currently merge the PR due to authorisation issues but there is currently looked into by @ilopmar and @alvarosanchez.
As soon as that is resolved there will be a snapshot available and we cut a release as well.

**Comment by donbeave** [comment]
@marceloverdijk ok, got it, sorry for pushing, I will be more patience )

**Comment by marceloverdijk** [comment]
No worries ☺️ sorry it is taking some time

**Review by ilopmar — APPROVED** [review]

**Comment by marceloverdijk** [comment]
@donbeave just to let you know but this is now included in Micronaut-GraphQL `2.1.0` being released last week.
Thx again for contributing this.

**Comment by donbeave** [comment]
@marceloverdijk yeah, I saw it. Thank you so much!
